File: scripts/import-interpreters.ts
import { drizzle } from "drizzle-orm/mysql2";
import { interpreters } from "../drizzle/schema";
import * as fs from "fs";
import * as path from "path";

const db = drizzle(process.env.DATABASE_URL!);

interface InterpreterRow {
  Language: string;
  Name: string;
  City: string;
  State: string;
  Phone: string;
  Email: string;
  Source: string;
}

async function importInterpreters() {
  console.log("Starting interpreter import...");
  
  // Read the combined CSV file
  const csvPath = path.join(process.cwd(), "../all_interpreters_combined_final.csv");
  
  if (!fs.existsSync(csvPath)) {
    console.error(`CSV file not found at: ${csvPath}`);
    process.exit(1);
  }

  const csvContent = fs.readFileSync(csvPath, "utf-8");
  const lines = csvContent.split("\n");
  const headers = lines[0]!.split(",").map(h => h.trim().replace(/"/g, ""));
  
  console.log(`Found ${lines.length - 1} rows to import`);
  console.log(`Headers: ${headers.join(", ")}`);

  const batchSize = 1000;
  let imported = 0;
  let batch: any[] = [];

  for (let i = 1; i < lines.length; i++) {
    const line = lines[i]!.trim();
    if (!line) continue;

    // Simple CSV parsing (handles basic cases)
    const values = line.split(",").map(v => v.trim().replace(/^"|"$/g, ""));
    
    const row: any = {};
    headers.forEach((header, idx) => {
      row[header] = values[idx] || "";
    });

    // Map metro from state if not present
    let metro = "";
    const state = row.State?.toUpperCase() || "";
    const city = row.City || "";
    
    // Basic metro mapping based on major cities
    if (city.toLowerCase().includes("tampa") || city.toLowerCase().includes("st. petersburg") || city.toLowerCase().includes("clearwater")) {
      metro = "Tampa-St. Petersburg-Clearwater, FL";
    } else if (city.toLowerCase().includes("miami") || city.toLowerCase().includes("fort lauderdale") || city.toLowerCase().includes("west palm")) {
      metro = "Miami-Fort Lauderdale-West Palm Beach, FL";
    } else if (city.toLowerCase().includes("orlando") || city.toLowerCase().includes("kissimmee")) {
      metro = "Orlando-Kissimmee-Sanford, FL";
    } else if (city.toLowerCase().includes("jacksonville")) {
      metro = "Jacksonville, FL";
    } else if (city.toLowerCase().includes("new york") || city.toLowerCase().includes("newark") || city.toLowerCase().includes("jersey city")) {
      metro = "New York-Newark-Jersey City, NY-NJ-PA";
    } else if (city.toLowerCase().includes("los angeles") || city.toLowerCase().includes("long beach") || city.toLowerCase().includes("anaheim")) {
      metro = "Los Angeles-Long Beach-Anaheim, CA";
    } else if (city.toLowerCase().includes("chicago") || city.toLowerCase().includes("naperville")) {
      metro = "Chicago-Naperville-Elgin, IL-IN-WI";
    } else if (city.toLowerCase().includes("houston") || city.toLowerCase().includes("woodlands") || city.toLowerCase().includes("sugar land")) {
      metro = "Houston-The Woodlands-Sugar Land, TX";
    } else if (city.toLowerCase().includes("phoenix") || city.toLowerCase().includes("mesa") || city.toLowerCase().includes("scottsdale")) {
      metro = "Phoenix-Mesa-Scottsdale, AZ";
    } else if (city.toLowerCase().includes("philadelphia") || city.toLowerCase().includes("camden") || city.toLowerCase().includes("wilmington")) {
      metro = "Philadelphia-Camden-Wilmington, PA-NJ-DE-MD";
    } else if (city.toLowerCase().includes("san antonio")) {
      metro = "San Antonio-New Braunfels, TX";
    } else if (city.toLowerCase().includes("san diego")) {
      metro = "San Diego-Carlsbad, CA";
    } else if (city.toLowerCase().includes("dallas") || city.toLowerCase().includes("fort worth") || city.toLowerCase().includes("arlington")) {
      metro = "Dallas-Fort Worth-Arlington, TX";
    } else if (city.toLowerCase().includes("san jose") || city.toLowerCase().includes("sunnyvale") || city.toLowerCase().includes("santa clara")) {
      metro = "San Jose-Sunnyvale-Santa Clara, CA";
    } else if (city.toLowerCase().includes("austin") || city.toLowerCase().includes("round rock")) {
      metro = "Austin-Round Rock, TX";
    } else if (city.toLowerCase().includes("seattle") || city.toLowerCase().includes("tacoma") || city.toLowerCase().includes("bellevue")) {
      metro = "Seattle-Tacoma-Bellevue, WA";
    } else if (city.toLowerCase().includes("denver") || city.toLowerCase().includes("aurora") || city.toLowerCase().includes("lakewood")) {
      metro = "Denver-Aurora-Lakewood, CO";
    } else if (city.toLowerCase().includes("washington") || city.toLowerCase().includes("arlington") || city.toLowerCase().includes("alexandria")) {
      metro = "Washington-Arlington-Alexandria, DC-VA-MD-WV";
    } else if (city.toLowerCase().includes("boston") || city.toLowerCase().includes("cambridge") || city.toLowerCase().includes("newton")) {
      metro = "Boston-Cambridge-Newton, MA-NH";
    } else if (city.toLowerCase().includes("atlanta") || city.toLowerCase().includes("sandy springs") || city.toLowerCase().includes("roswell")) {
      metro = "Atlanta-Sandy Springs-Roswell, GA";
    } else if (city.toLowerCase().includes("portland") && (state === "OR" || state === "WA")) {
      metro = "Portland-Vancouver-Hillsboro, OR-WA";
    } else if (city.toLowerCase().includes("baltimore") || city.toLowerCase().includes("columbia") || city.toLowerCase().includes("towson")) {
      metro = "Baltimore-Columbia-Towson, MD";
    }

    batch.push({
      name: row.Name || "Unknown",
      language: row.Language || "Unknown",
      city: row.City || null,
      state: row.State || null,
      metro: metro || null,
      phone: row.Phone || null,
      email: row.Email || null,
      source: row.Source || "Unknown",
      certifications: null,
      specialties: null,
      latitude: null,
      longitude: null,
    });

    if (batch.length >= batchSize) {
      try {
        await db.insert(interpreters).values(batch);
        imported += batch.length;
        console.log(`Imported ${imported} interpreters...`);
        batch = [];
      } catch (error) {
        console.error("Error importing batch:", error);
      }
    }
  }

  // Import remaining batch
  if (batch.length > 0) {
    try {
      await db.insert(interpreters).values(batch);
      imported += batch.length;
      console.log(`Imported ${imported} interpreters...`);
    } catch (error) {
      console.error("Error importing final batch:", error);
    }
  }

  console.log(`\n✓ Import complete! Total interpreters imported: ${imported}`);
  process.exit(0);
}

importInterpreters().catch((error) => {
  console.error("Fatal error during import:", error);
  process.exit(1);
});

File: scripts/import_interpreters.py
#!/usr/bin/env python3
"""
Fast bulk import of interpreters using Python and MySQL connector
"""
import csv
import os
import mysql.connector
from urllib.parse import urlparse

def parse_database_url(url):
    """Parse DATABASE_URL into connection parameters"""
    parsed = urlparse(url)
    return {
        'host': parsed.hostname,
        'port': parsed.port or 3306,
        'user': parsed.username,
        'password': parsed.password,
        'database': parsed.path[1:] if parsed.path else None
    }

def map_metro(city, state):
    """Map city and state to metro area"""
    if not city:
        return None
    
    city_lower = city.lower()
    state_upper = state.upper() if state else ""
    
    metro_mappings = {
        ("tampa", "st. petersburg", "clearwater"): "Tampa-St. Petersburg-Clearwater, FL",
        ("miami", "fort lauderdale", "west palm"): "Miami-Fort Lauderdale-West Palm Beach, FL",
        ("orlando", "kissimmee"): "Orlando-Kissimmee-Sanford, FL",
        ("jacksonville",): "Jacksonville, FL",
        ("new york", "newark", "jersey city"): "New York-Newark-Jersey City, NY-NJ-PA",
        ("los angeles", "long beach", "anaheim"): "Los Angeles-Long Beach-Anaheim, CA",
        ("chicago", "naperville"): "Chicago-Naperville-Elgin, IL-IN-WI",
        ("houston", "woodlands", "sugar land"): "Houston-The Woodlands-Sugar Land, TX",
        ("phoenix", "mesa", "scottsdale"): "Phoenix-Mesa-Scottsdale, AZ",
        ("philadelphia", "camden", "wilmington"): "Philadelphia-Camden-Wilmington, PA-NJ-DE-MD",
        ("san antonio",): "San Antonio-New Braunfels, TX",
        ("san diego",): "San Diego-Carlsbad, CA",
        ("dallas", "fort worth", "arlington"): "Dallas-Fort Worth-Arlington, TX",
        ("san jose", "sunnyvale", "santa clara"): "San Jose-Sunnyvale-Santa Clara, CA",
        ("austin", "round rock"): "Austin-Round Rock, TX",
        ("seattle", "tacoma", "bellevue"): "Seattle-Tacoma-Bellevue, WA",
        ("denver", "aurora", "lakewood"): "Denver-Aurora-Lakewood, CO",
        ("washington", "arlington", "alexandria"): "Washington-Arlington-Alexandria, DC-VA-MD-WV",
        ("boston", "cambridge", "newton"): "Boston-Cambridge-Newton, MA-NH",
        ("atlanta", "sandy springs", "roswell"): "Atlanta-Sandy Springs-Roswell, GA",
        ("baltimore", "columbia", "towson"): "Baltimore-Columbia-Towson, MD",
    }
    
    for keywords, metro in metro_mappings.items():
        if any(keyword in city_lower for keyword in keywords):
            return metro
    
    if "portland" in city_lower and state_upper in ("OR", "WA"):
        return "Portland-Vancouver-Hillsboro, OR-WA"
    
    return None

def main():
    print("Starting fast interpreter import...")
    
    # Get database connection
    database_url = os.environ.get('DATABASE_URL')
    if not database_url:
        print("ERROR: DATABASE_URL not set")
        return
    
    conn_params = parse_database_url(database_url)
    print(f"Connecting to database: {conn_params['host']}")
    
    conn = mysql.connector.connect(**conn_params)
    cursor = conn.cursor()
    
    # Read CSV
    csv_path = "/home/ubuntu/all_interpreters_combined_final.csv"
    if not os.path.exists(csv_path):
        print(f"ERROR: CSV file not found at {csv_path}")
        return
    
    print(f"Reading CSV from {csv_path}...")
    
    with open(csv_path, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        
        batch = []
        batch_size = 1000
        imported = 0
        
        for row in reader:
            metro = map_metro(row.get('City', ''), row.get('State', ''))
            
            batch.append((
                row.get('Name', 'Unknown'),
                row.get('Language', 'Unknown'),
                row.get('City') or None,
                row.get('State') or None,
                metro,
                row.get('Phone') or None,
                row.get('Email') or None,
                row.get('Source', 'Unknown'),
                None,  # certifications
                None,  # specialties
                None,  # latitude
                None,  # longitude
            ))
            
            if len(batch) >= batch_size:
                cursor.executemany("""
                    INSERT INTO interpreters 
                    (name, language, city, state, metro, phone, email, source, 
                     certifications, specialties, latitude, longitude)
                    VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                """, batch)
                conn.commit()
                imported += len(batch)
                print(f"Imported {imported} interpreters...")
                batch = []
        
        # Import remaining
        if batch:
            cursor.executemany("""
                INSERT INTO interpreters 
                (name, language, city, state, metro, phone, email, source, 
                 certifications, specialties, latitude, longitude)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """, batch)
            conn.commit()
            imported += len(batch)
            print(f"Imported {imported} interpreters...")
    
    cursor.close()
    conn.close()
    
    print(f"\n✓ Import complete! Total: {imported} interpreters")

if __name__ == "__main__":
    main()
